solver/src/solver/http_solver.rs

```diff
@@ -512,43 +294,191 @@ impl HttpSolver {
+        self.cost_for_gas(*GAS_PER_UNISWAP)
+    }
+
+    fn balancer_cost(&self) -> CostModel {
+        self.cost_for_gas(*GAS_PER_BALANCER_SWAP)
+    }
+
+    fn order_fee(&self, order: &LimitOrder) -> FeeModel {
+        let ceiled_div = (order.fee_amount.to_f64_lossy() / self.fee_factor).ceil();
+        FeeModel {
+            amount: U256::from_f64_lossy(ceiled_div),
+            token: order.sell_token,
         }
     }
-    (
-        constant_product_orders,
-        weighted_product_orders,
-        stable_pool_orders,
-    )
 }
 
-// TODO: This currently does **NOT** consider balancer pools, but definitely should.
-fn remove_orders_without_native_connection(
-    orders: &mut Vec<LimitOrder>,
-    amms: &[ConstantProductOrder],
-    native_token: &H160,
-) {
+fn token_models(
+    token_infos: &HashMap<H160, TokenInfo>,
+    price_estimates: &HashMap<H160, f64>,
+    buffers: &HashMap<H160, U256>,
+    gas_model: &GasModel,
+) -> BTreeMap<H160, TokenInfoModel> {
+    token_infos
+        .iter()
+        .map(|(address, token_info)| {
+            let external_price = match price_estimates.get(address).copied() {
+                Some(price) if price.is_finite() => Some(price),
+                _ => None,
+            };
+            (
+                *address,
+                TokenInfoModel {
+                    decimals: token_info.decimals,
+                    external_price,
+                    normalize_priority: Some(if &gas_model.native_token == address {
+                        1
+                    } else {
+                        0
+                    }),
+                    internal_buffer: buffers.get(address).copied(),
+                },
+            )
+        })
+        .collect()
+}
+
+fn order_models(
+    orders: &[LimitOrder],
+    fee_connected_tokens: &HashSet<H160>,
+    gas_model: &GasModel,
+) -> BTreeMap<usize, OrderModel> {
+    orders
+        .iter()
+        .enumerate()
+        .filter_map(|(index, order)| {
+            if ![order.sell_token, order.buy_token]
+                .iter()
+                .any(|token| fee_connected_tokens.contains(token))
+            {
+                return None;
+            }
+
+            Some((
+                index,
+                OrderModel {
+                    sell_token: order.sell_token,
+                    buy_token: order.buy_token,
+                    sell_amount: order.sell_amount,
+                    buy_amount: order.buy_amount,
+                    allow_partial_fill: order.partially_fillable,
+                    is_sell_order: matches!(order.kind, OrderKind::Sell),
+                    fee: gas_model.order_fee(order),
+                    cost: gas_model.order_cost(),
+                    is_liquidity_order: order.is_liquidity_order,
+                },
+            ))
+        })
+        .collect()
+}
+
+fn amm_models(liquidity: &[Liquidity], gas_model: &GasModel) -> BTreeMap<usize, AmmModel> {
```

Comment:
:meat_on_bone: of the PR, no longer requires splitting `Liquidity` into separate collections for each liquidity kind.